### PR TITLE
Add new 'uPortal-session' submodule which adds support for session cl…

### PR DESCRIPTION
**Pull request by groybal**
…ustering / replication / failover.

##### Checklist

-   [x] the [individual contributor license agreement][] is signed
-   [x] commit message follows [commit guidelines][]
-   [x] tests are included
-   [x] documentation is changed or added
-   [x] any changes that impact configuration or the database DDL added to `CHANGES.md`

##### Description of change

A new uPortal-session submodule was added, which uses Spring Session in order to store the web sessions externally in Redis.  This is disabled by default, but can be enabled with an environment variable or system property.  The Redis connection can be configured for standalone, cluster, or sentinel modes.

[individual contributor license agreement]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#individual-contributor-license-agreement
[commit guidelines]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#commit
[message properties]: https://github.com/Jasig/uPortal/tree/master/uportal-war/src/main/resources/properties/i18n
[WCAG 2.0 AA]: https://www.w3.org/WAI/WCAG20/quickref/?levels=aaa&technologies=smil%2Cpdf%2Cflash%2Csl

### CODE REVIEW

**Comment by jgribonvald** [comment]
Nice starting work !

Also you should watch on the CAS lib to link the session to the CAS ticket to be able to interpret the CAS logout (not lib for CAS in a distributed context exist for with redis but it won't be too much to do)

Also for working with redis in a HA and fail-over context I would advice to use redis (cache mode and not database) + sentinel. And using HAProxy could be powerfull for optimizing where to make read of write request (https://www.willandskill.se/sv/articles/setup-a-highly-available-redis-cluster-with-sentinel-and-haproxy)

For code examples I've pushed some years ago a conf + unit test on CAS around that, watch there if it can help: https://github.com/apereo/cas/commit/171cf34089584e0c5efe2e4a412efa76f37dbc72 and follow the history of change around https://github.com/apereo/cas/commits/master/support/cas-server-support-redis-core/src/main/java/org/apereo/cas/redis/core/RedisObjectFactory.java

After that you would be able to put the cache into redis too and it will consume really less server ressources ;)

**Comment by bjagg** [comment]
@jgribonvald could we break out your suggestions into Issues that can be tackled at a later date? We just want to get the basic session replication support in before our next release. 

Anything you think is really needed for this PR?

**Comment by jgribonvald** [comment]
> @jgribonvald could we break out your suggestions into Issues that can be tackled at a later date? We just want to get the basic session replication support in before our next release. 
> 
> Anything you think is really needed for this PR?

Don't worry 😉 it's OK for a first introduction of the feature.

But maybe one thing is missing, the test ?

File: uPortal-session/src/main/java/org/apereo/portal/session/redis/RedisSessionConfig.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        if (this.redisClusterMaxRedirects != null) {
+            clusterConfig.setMaxRedirects(this.redisClusterMaxRedirects);
+        }
+        return new JedisConnectionFactory(clusterConfig);
+    }
+
+    private JedisConnectionFactory createBaseSentinelConnectionFactory() {
+        RedisSentinelConfiguration sentinelConfig = new RedisSentinelConfiguration();
+        sentinelConfig.master(this.redisSentinelMaster);
+        String[] nodesArray = this.redisSentinelNodes.split(",");
```
Comment (ChristianMurphy) [inline uPortal-session/src/main/java/org/apereo/portal/session/redis/RedisSessionConfig.java:97]:
```
D:\a\uPortal\uPortal\uPortal-session\src\main\java\org\apereo\portal\session\redis\RedisSessionConfig.java:97: warning: [StringSplitter] String.split(String) has surprising behavior

        String[] nodesArray = this.redisSentinelNodes.split(",");
                                                           ^
    (see https://errorprone.info/bugpattern/StringSplitter)
  Did you mean 'Iterable<String> nodesArray = Splitter.on(',').split(this.redisSentinelNodes);'?
```

https://github.com/uPortal-Project/uPortal/actions/runs/6752494006/job/18357912230?pr=2699#step:5:4685

File: uPortal-session/src/main/java/org/apereo/portal/session/redis/RedisSessionConfig.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        for (String node : nodesArray) {
+            String[] hostAndPort = node.split(":");
```
Comment (ChristianMurphy) [inline uPortal-session/src/main/java/org/apereo/portal/session/redis/RedisSessionConfig.java:99]:
```
D:\a\uPortal\uPortal\uPortal-session\src\main\java\org\apereo\portal\session\redis\RedisSessionConfig.java:99: warning: [StringSplitter] String.split(String) has surprising behavior
            String[] hostAndPort = node.split(":");
                                             ^
    (see https://errorprone.info/bugpattern/StringSplitter)
  Did you mean 'List<String> hostAndPort = Splitter.on(':').splitToList(node);'?

```

https://github.com/uPortal-Project/uPortal/actions/runs/6752494006/job/18357912230?pr=2699#step:5:4691

File: uPortal-session/src/test/java/org/apereo/portal/session/redis/RedisSessionInitializerTest.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    @Test
+    public void testOnStartupAddsFilterWhenStoreTypeEnvironmentVariableSetToRedis()
+            throws Exception {
+        try {
+            this.setEnvironmentVariableValue("ORG_APEREO_PORTAL_SESSION_STORETYPE", "redis");
+            this.redisSessionInitializer.onStartup(this.servletContext);
+
+            verify(this.servletContext, times(1)).addFilter(anyString(), any(Filter.class));
```
Comment (ChristianMurphy) [inline uPortal-session/src/test/java/org/apereo/portal/session/redis/RedisSessionInitializerTest.java:88]:
```
> Task :uPortal-session:test FAILED

org.apereo.portal.session.redis.RedisSessionInitializerTest > testOnStartupAddsFilterWhenStoreTypeEnvironmentVariableSetToRedis FAILED
    org.mockito.exceptions.verification.WantedButNotInvoked at RedisSessionInitializerTest.java:88
```

https://github.com/uPortal-Project/uPortal/actions/runs/6752494006/job/18357912230?pr=2699#step:5:4719

File: uPortal-session/src/main/java/org/apereo/portal/session/redis/RedisSessionConfig.java (hunk 1 above)
Comment (groybal) [reply uPortal-session/src/main/java/org/apereo/portal/session/redis/RedisSessionConfig.java]: fixed

File: uPortal-session/src/main/java/org/apereo/portal/session/redis/RedisSessionConfig.java (hunk 2 above)
Comment (groybal) [reply uPortal-session/src/main/java/org/apereo/portal/session/redis/RedisSessionConfig.java]: fixed

File: uPortal-session/src/test/java/org/apereo/portal/session/redis/RedisSessionInitializerTest.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        }
+    }
+
+    @Test
+    public void testOnStartupAddsFilterWhenStoreTypeEnvironmentVariableSetToRedis()
+            throws Exception {
+        try {
+            this.setEnvironmentVariableValue("ORG_APEREO_PORTAL_SESSION_STORETYPE", "redis");
+            this.redisSessionInitializer.onStartup(this.servletContext);
+            verify(this.servletContext, times(1))
```
Comment (ChristianMurphy) [inline uPortal-session/src/test/java/org/apereo/portal/session/redis/RedisSessionInitializerTest.java:90]:
```
> Task :uPortal-session:test FAILED

org.apereo.portal.session.redis.RedisSessionInitializerTest > testOnStartupAddsFilterWhenStoreTypeEnvironmentVariableSetToRedis FAILED
    org.mockito.exceptions.verification.WantedButNotInvoked at RedisSessionInitializerTest.java:91
```

**Review by bjagg — APPROVED** [review]
🎖️

**Pull request merged**